jlm/llvm/ir/cfg.hpp

```diff
@@ -156,111 +156,10 @@ class ExitNode final : public ControlFlowGraphNode
 
 class ControlFlowGraph final
 {
-  class iterator final
-  {
-  public:
-    inline iterator(std::unordered_set<std::unique_ptr<BasicBlock>>::iterator it)
-        : it_(it)
-    {}
-
-    inline bool
-    operator==(const iterator & other) const noexcept
-    {
-      return it_ == other.it_;
-    }
-
-    inline bool
-    operator!=(const iterator & other) const noexcept
-    {
-      return !(*this == other);
-    }
-
-    inline const iterator &
-    operator++() noexcept
-    {
-      ++it_;
-      return *this;
-    }
-
-    inline const iterator
-    operator++(int) noexcept
-    {
-      iterator tmp(it_);
-      it_++;
-      return tmp;
-    }
-
-    inline BasicBlock *
-    node() const noexcept
-    {
-      return it_->get();
-    }
-
-    inline BasicBlock &
-    operator*() const noexcept
-    {
-      return *it_->get();
-    }
-
-    inline BasicBlock *
-    operator->() const noexcept
-    {
-      return node();
-    }
-
-  private:
-    std::unordered_set<std::unique_ptr<BasicBlock>>::iterator it_;
-  };
-
-  class const_iterator final
-  {
-  public:
-    inline const_iterator(std::unordered_set<std::unique_ptr<BasicBlock>>::const_iterator it)
-        : it_(it)
-    {}
-
-    inline bool
-    operator==(const const_iterator & other) const noexcept
-    {
-      return it_ == other.it_;
-    }
-
-    inline bool
-    operator!=(const const_iterator & other) const noexcept
-    {
-      return !(*this == other);
-    }
-
-    inline const const_iterator &
-    operator++() noexcept
-    {
-      ++it_;
-      return *this;
-    }
-
-    inline const const_iterator
-    operator++(int) noexcept
-    {
-      const_iterator tmp(it_);
-      it_++;
-      return tmp;
-    }
-
-    inline const BasicBlock &
-    operator*() noexcept
-    {
-      return *it_->get();
-    }
-
-    inline const BasicBlock *
-    operator->() noexcept
-    {
-      return it_->get();
-    }
-
-  private:
-    std::unordered_set<std::unique_ptr<BasicBlock>>::const_iterator it_;
-  };
+  using iterator =
+      util::PtrIterator<BasicBlock, std::unordered_set<std::unique_ptr<BasicBlock>>::iterator>;
+  using const_iterator = util::
+      PtrIterator<BasicBlock, std::unordered_set<std::unique_ptr<BasicBlock>>::const_iterator>;
```

Comment:
This is not how you make a const iterator.

When we use `std::unordered_set<std::unique_ptr<BasicBlock>>::const_iterator`, that constness is about not being able to change what the `unique_ptr` is pointing to.
The iterator never exposes the `unique_ptr` anyways, so it is impossible to re-assign it.

To make a const iterator over `const BasicBlock &`, specify `const BasicBlock` as the type.

Reply:
Added a const